File: core/FindKThSmallestPairDistance/FindKThSmallestPairDistance.java
package com.dbc;

import java.util.Arrays;

public class FindKThSmallestPairDistance {
    public int smallestDistancePair(int[] nums, int k) {
        Arrays.sort(nums);
        int left = 0, right = nums[nums.length - 1] - nums[0];
        while (left <= right) {
            int mid = (left + right) / 2;
            int count = 0;
            for (int i = 0, j = 0; j < nums.length; j++) {
                while (nums[j] - nums[i] > mid) i++;
                count += j - i;
            }
            if (count >= k) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}

File: core/FindKThSmallestPairDistance/FindKThSmallestPairDistance.py
from typing import List
from bisect import bisect_left


def smallestDistancePair(nums: List[int], k: int) -> int:
    nums.sort()
    left, right = 0, nums[-1] - nums[0]
    while left <= right:
        mid = (left + right) // 2
        point, count = 0, 0
        for index, num in enumerate(nums):
            while num - nums[point] > mid:
                point += 1
            count += index - point
        if count >= k:
            right = mid - 1
        else:
            left = mid + 1

    return left






